server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -271,15 +274,36 @@ private TopDocs searchOperations(ScoreDoc after) throws IOException {
                 }
                 // TODO: pass the latest timestamp from engine.
                 final long autoGeneratedIdTimestamp = -1;
+                // Add the parent id if present
+                String parent = getParentDocValue(mapperService, leaf.reader(), type, segmentDocID);
                 op = new Translog.Index(type, id, seqNo, primaryTerm, version, VersionType.EXTERNAL,
-                    source.toBytesRef().bytes, fields.routing(), null, autoGeneratedIdTimestamp);
+                    source.toBytesRef().bytes, fields.routing(), parent, autoGeneratedIdTimestamp);
             }
         }
         assert fromSeqNo <= op.seqNo() && op.seqNo() <= toSeqNo && lastSeenSeqNo < op.seqNo() : "Unexpected operation; " +
             "last_seen_seqno [" + lastSeenSeqNo + "], from_seqno [" + fromSeqNo + "], to_seqno [" + toSeqNo + "], op [" + op + "]";
         return op;
     }
 
+    /**
+     * Return the parent id from the {@link LeafReader} or null if the {@link ParentFieldMapper} is
+     * not active.
+     */
+    private String getParentDocValue(MapperService mapperService, LeafReader leafReader,
+                                        String type, int segmentDocId) throws IOException {
+        ParentFieldMapper parentFieldMapper = mapperService.documentMapper(type).parentFieldMapper();
+        if (parentFieldMapper.active() == false) {
+            return null;
+        }
+        SortedDocValues dvs = leafReader.getSortedDocValues(parentFieldMapper.name());
```

Comment:
This class requires the index version >= 6.5 which should have a single type. Can we get the mapping type directly from MapperService and fold this DocValues into the ParallelArray?